From 64debbaf5bdef93302f29c3f6cd395e3e22d65d8 Mon Sep 17 00:00:00 2001
From: Luca Weiss <luca@z3ntu.xyz>
Date: Wed, 28 Oct 2020 23:05:52 +0100
Subject: [PATCH] utils: fix missing newline on new shell entries

Previously if the existing file didn't end with a newline then two shell
entries would be written without any separator to the file leading to a
broken file.
---
 src/utils.c | 17 +++++++++++++++++
 1 file changed, 17 insertions(+)

diff --git a/src/utils.c b/src/utils.c
index d61274a..041a087 100644
--- a/src/utils.c
+++ b/src/utils.c
@@ -488,6 +488,23 @@ shell_parser_set_variable (ShellParser *parser,
         ret = TRUE;
     } else {
         if (add_if_unset) {
+            GList *previous;
+            struct ShellEntry *previous_entry = NULL;
+
+            /* We need to add a separator (\n) between two items if there isn't one already. */
+            if ((previous = g_list_last(parser->entry_list)) != NULL) {
+                previous_entry = previous->data;
+            }
+            if (previous_entry && previous_entry->type != SHELL_ENTRY_TYPE_SEPARATOR) {
+                struct ShellEntry *separator_entry;
+
+                g_debug ("Adding separator entry");
+                separator_entry = g_new0 (struct ShellEntry, 1);
+                separator_entry->type = SHELL_ENTRY_TYPE_SEPARATOR;
+                separator_entry->string = g_strdup ("\n");
+                parser->entry_list = g_list_append (parser->entry_list, separator_entry);
+            }
+
             found_entry = g_new0 (struct ShellEntry, 1);
             found_entry->type = SHELL_ENTRY_TYPE_ASSIGNMENT;
             found_entry->variable = g_strdup (variable);
-- 
2.25.1

